src/csexp_rpc/csexp_rpc.ml

```diff
@@ -15,6 +15,73 @@ end
 
 module Session_id = Id.Make ()
 
+module Socket = struct
+  module type Unix_socket = sig
+    val connect : Unix.file_descr -> socket:string -> unit
+
+    val bind : Unix.file_descr -> socket:string -> unit
+  end
+
+  module U = Unix
+
+  module Mac : Unix_socket = struct
+    external pthread_chdir : string -> unit = "dune_pthread_chdir" [@@noalloc]
+
+    let with_chdir fd ~socket ~f =
+      let old = Sys.getcwd () in
+      let dir = Filename.dirname socket in
+      let sock = Filename.basename socket in
+      pthread_chdir dir;
+      Exn.protectx (Unix.ADDR_UNIX sock) ~f:(f fd) ~finally:(fun _ ->
+          pthread_chdir old)
+
+    let connect fd ~socket : unit = with_chdir fd ~socket ~f:Unix.connect
+
+    let bind fd ~socket : unit = with_chdir fd ~socket ~f:Unix.bind
+  end
+
+  module Unix : Unix_socket = struct
+    let addr socket =
+      Unix.ADDR_UNIX
+        (match
+           let cwd = Sys.getcwd () in
+           String.drop_prefix socket ~prefix:(cwd ^ "/")
+         with
+        | Some s -> "./" ^ s
+        | None -> socket)
+
+    let connect fd ~socket = Unix.connect fd (addr socket)
+
+    let bind fd ~socket = Unix.bind fd (addr socket)
+  end
+
+  module Fail : Unix_socket = struct
+    let connect _ ~socket:_ = Code_error.raise "Fail.connect" []
+
+    let bind _ ~socket:_ = Code_error.raise "Fail.bind" []
+  end
+
+  external is_osx : unit -> bool = "dune_pthread_chdir_is_osx" [@@noalloc]
+
+  module Sel = (val if is_osx () then
+                      (module Mac)
+                    else if Sys.unix then
+                      (module Unix)
+                    else
+                      (module Fail) : Unix_socket)
```

Comment:
What about Windows? I forgot what we do on Windows.

Reply:
We don't use Unix domain sockets on windows so the problem does not exist there.